Source/TimeIntegration/ROMSX_advance_2d.cpp

```diff
@@ -284,6 +288,205 @@ ROMSX::advance_2d (int lev,
+        auto fomn=fab_fomn.array();
+        auto om_u=fab_om_u.array();
+        auto on_v=fab_on_v.array();
+        auto om_r=fab_om_r.array();
+        auto on_r=fab_on_r.array();
+        auto om_p=fab_om_p.array();
+        auto on_p=fab_on_p.array();
+
+        auto Drhs=fab_Drhs.array();
+        auto Drhs_p=fab_Drhs_p.array();
+        auto Dnew=fab_Dnew.array();
+        auto zwrk=fab_zwrk.array();
+        auto gzeta=fab_gzeta.array();
+        auto gzeta2=fab_gzeta2.array();
+        auto gzetaSA=fab_gzetaSA.array();
+        auto Dstp=fab_Dstp.array();
+        auto DUon=fab_DUon.array();
+        auto DVom=fab_DVom.array();
+        auto rhs_ubar=fab_rhs_ubar.array();
+        auto rhs_vbar=fab_rhs_vbar.array();
+        auto rhs_zeta=fab_rhs_zeta.array();
+        auto zeta_new=fab_zeta_new.array();
+
+        auto weight1 = vec_weight1.dataPtr();
+        auto weight2 = vec_weight2.dataPtr();
+       if ((verbose > 2) && predictor_2d_step && my_iif == 0) {
+           amrex::PrintToFile("ru_startadvance").SetPrecision(18)<<FArrayBox(ru)<<std::endl;
+           amrex::PrintToFile("rv_startadvance").SetPrecision(18)<<FArrayBox(rv)<<std::endl;
+           amrex::PrintToFile("u_startadvance2").SetPrecision(18)<<FArrayBox(u)<<std::endl;
+           amrex::PrintToFile("v_startadvance2").SetPrecision(18)<<FArrayBox(v)<<std::endl;
+       }
+
+        //From ana_grid.h and metrics.F
+        amrex::ParallelFor(tbxp2,
+        [=] AMREX_GPU_DEVICE (int i, int j, int)
+        {
+              pm(i,j,0)=dxi[0];
+              pn(i,j,0)=dxi[1];
+              rhs_ubar(i,j,0)=0.0;
+              rhs_vbar(i,j,0)=0.0;
+        });
+
+        amrex::ParallelFor(tbxp2,
+        [=] AMREX_GPU_DEVICE (int i, int j, int k)
+        {
+              //Note: are the comment definitons right? Don't seem to match metrics.f90
+              om_v(i,j,0)=1.0/dxi[0]; // 2/(pm(i,j-1)+pm(i,j))
+              on_u(i,j,0)=1.0/dxi[1]; // 2/(pm(i,j-1)+pm(i,j))
+              om_r(i,j,0)=1.0/dxi[0]; // 1/pm(i,j)
+              on_r(i,j,0)=1.0/dxi[1]; // 1/pn(i,j)
+              //todo: om_p on_p
+              om_p(i,j,0)=1.0/dxi[0]; // 4/(pm(i-1,j-1)+pm(i-1,j)+pm(i,j-1)+pm(i,j))
+              on_p(i,j,0)=1.0/dxi[1]; // 4/(pn(i-1,j-1)+pn(i-1,j)+pn(i,j-1)+pn(i,j))
+              on_v(i,j,0)=1.0/dxi[1]; // 2/(pn(i-1,j)+pn(i,j))
+              om_u(i,j,0)=1.0/dxi[0]; // 2/(pm(i-1,j)+pm(i,j))
+        });
+        amrex::ParallelFor(tbxp2,
+        [=] AMREX_GPU_DEVICE (int i, int j, int  )
+        {
+
+              const auto prob_lo         = geomdata.ProbLo();
+              const auto dx              = geomdata.CellSize();
+
+              pm(i,j,0)=dxi[0];
+              pn(i,j,0)=dxi[1];
+              //defined UPWELLING
+              Real f0=-8.26e-5;
+              Real beta=0.0;
+              Real Esize=1000*(Mm);
+              Real y = prob_lo[1] + (j + 0.5) * dx[1];
+              Real f=fomn(i,j,0)=f0+beta*(y-.5*Esize);
```

Comment:
```suggestion
              Real f=f0+beta*(y-.5*Esize);
```

Reply:
This pattern appears in a number of places. I will submit a PR tomorrow. I did not change it last time because if the Box is 2D, this is harmless even though it wastes a write. But if it is not 2d, there are race conditions.